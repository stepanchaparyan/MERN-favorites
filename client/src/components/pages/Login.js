import React, { useState, useContext, useEffect } from 'react';
import { Link } from 'react-router-dom';
import AuthContext from '../../context/authContext/authContext';

const Login = props => {
  const { login, isAuthencated, error, clearErrors } = useContext(AuthContext);
  useEffect(() => {
    if (isAuthencated) {
      props.history.push('/');
      clearErrors();
    } else {
      clearErrors();
    }
  }, [isAuthencated, props.history]);
  const [user, setUser] = useState({
    email: '',
    password: ''
  });
  const { email, password } = user;

  const onchange = e => {
    setUser({
      ...user,
      [e.target.name]: e.target.value
    });
    if (error !== null) {
      clearErrors();
    }
  };
  const onsubmit = e => {
    e.preventDefault();
    login({
      email,
      password
    });
    clearErrors();
  };
  return (
    <div className='login'>
      <h1>Login</h1>
      <form onSubmit={onsubmit}>
        <input
          type='email'
          name='email'
          placeholder='Email'
          value={email}
          onChange={onchange}
        />
        <input
          type='password'
          name='password'
          placeholder='Password'
          value={password}
          onChange={onchange}
          required
        />
        <input type='submit' value='Login' className='btn' />
      </form>
      <div className='question'>
        {error !== null && (
          <button className='danger' type='button'>
            {error} <span onClick={() => clearErrors()}>X</span>
          </button>
        )}
        <p>
          Dont' have an accout? <Link to='/register'>Sign Up</Link>
        </p>
      </div>
    </div>
  );
};
export default Login;
