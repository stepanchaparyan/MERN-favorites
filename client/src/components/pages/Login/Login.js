import React, { useState, useContext, useEffect } from 'react';
import { useIntl } from 'react-intl';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import AuthContext from '../../../context/authContext/authContext';
import {
  Container,
  Title,
  Form,
  Input,
  LoginButton,
  Errors,
  QuestionText,
  ErrorButton
} from './LoginStyled';
import localization from './localization';
import { FORM, LINK } from '../../../constants';

const { INPUT } = FORM;

const Login = props => {
  const { login, isAuthencated, error, clearErrors } = useContext(AuthContext);
  const { formatMessage } = useIntl();

  useEffect(() => {
    if (isAuthencated) {
      props.history.push(LINK.TO.WELCOME);
      clearErrors();
    } else {
      clearErrors();
    }
  }, [isAuthencated, props.history]);

  const [user, setUser] = useState({
    email: '',
    password: ''
  });
  const { email, password } = user;

  const onchange = e => {
    setUser({
      ...user,
      [e.target.name]: e.target.value
    });
    if (error !== null) {
      clearErrors();
    }
  };
  const onsubmit = e => {
    e.preventDefault();
    login({
      email,
      password
    });
    clearErrors();
  };

  return (
    <Container>
      <Title>{formatMessage(localization.login)}</Title>
      <Form onSubmit={e => onsubmit(e)}>
        <Input
          type={INPUT.TYPE.EMAIL}
          name={INPUT.NAME.EMAIL}
          placeholder={formatMessage(localization.email)}
          value={email}
          onChange={e => onchange(e)}
          required
        />
        <Input
          type={INPUT.TYPE.PASSWORD}
          name={INPUT.NAME.PASSWORD}
          placeholder={formatMessage(localization.password)}
          value={password}
          onChange={e => onchange(e)}
          required
        />
        <LoginButton>{formatMessage(localization.login)}</LoginButton>
      </Form>
      {error !== null && (
        <Errors>
          {!Array.isArray(error) ? (
            <ErrorButton type={INPUT.TYPE.BUTTON} onClick={() => clearErrors()}>
              {error}
            </ErrorButton>
          ) : (
            error.map(err => (
              <ErrorButton key={err.msg} type={INPUT.TYPE.BUTTON} onClick={() => clearErrors()}>
                {err.msg}
              </ErrorButton>
            ))
          )}
        </Errors>
      )}
      <QuestionText>
        {formatMessage(localization.dontHaveAnAccout)} &nbsp;
        <Link to={LINK.TO.REGISTER}>{formatMessage(localization.signup)}</Link>
      </QuestionText>
    </Container>
  );
};

Login.propTypes = {
  history: PropTypes.object
};

export default Login;
