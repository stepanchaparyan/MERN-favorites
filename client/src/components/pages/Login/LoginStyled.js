import styled from 'styled-components';

export const Container = styled.div`
  display: flex;
  justify-content: center;
  flex-direction: column;
  color: ${props => props.theme.cadetblue};
`;

export const Title = styled.div`
  display: flex;
  justify-content: center;
  padding: 64px 32px 32px;
  font-weight: 600;
  font-size: 30px;
`;

export const Form = styled.form`
  display: flex;
  flex-direction: column;
  align-items: center;
`;

export const Input = styled.input`
  border: 0px;
  border-bottom: 1px solid ${props => props.theme.cadetblue};
  width: 30%;
  padding: 7px 10px;
  margin: 2px;
  outline: none;

  &[type='submit'] {
    width: 20%;
    border: 0;
    background-color: ${props => props.theme.cadetblue};
    font-weight: 600;
    font-size: 16px;
    margin-top: 20px;
    border-radius: 3%;
    cursor: pointer;
  }
`;

export const Errors = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  margin: 10px;
`;

export const QuestionText = styled.div`
  display: flex;
  align-self: center;
  margin: 10px;
  & > a {
    text-decoration: none;
    color: black;
    font-weight: 600;
  }
`;

export const ErrorButton = styled.button`
  color: red;
  background-color: transparent;
  border: none;
  outline: none;
  cursor: pointer;
`;
