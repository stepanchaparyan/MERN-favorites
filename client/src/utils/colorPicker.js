export const colorPicker = category => {
  let color;
  switch (category) {
    case 'Film':
      color = 'red';
      break;
    case 'Music':
      color = 'blue';
      break;
    case 'Books':
      color = 'limeGreen';
      break;
    case 'Other':
      color = 'orange';
      break;
    default:
      color = 'black';
  }
  return color;
};
