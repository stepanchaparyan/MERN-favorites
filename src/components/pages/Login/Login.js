import React, { useState, useContext, useEffect } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import AuthContext from '../../../context/authContext/authContext';
import {
  Container,
  Title,
  Form,
  Input,
  Errors,
  QuestionText,
  ErrorButton
} from './LoginStyled';

const Login = props => {
  const { login, isAuthencated, error, clearErrors } = useContext(AuthContext);
  useEffect(() => {
    if (isAuthencated) {
      props.history.push('/');
      clearErrors();
    } else {
      clearErrors();
    }
  }, [isAuthencated, props.history]);

  const [user, setUser] = useState({
    email: '',
    password: ''
  });
  const { email, password } = user;

  const onchange = e => {
    setUser({
      ...user,
      [e.target.name]: e.target.value
    });
    if (error !== null) {
      clearErrors();
    }
  };
  const onsubmit = e => {
    e.preventDefault();
    login({
      email,
      password
    });
    clearErrors();
  };

  return (
    <Container>
      <Title>Login</Title>
      <Form onSubmit={onsubmit}>
        <Input
          type='email'
          name='email'
          placeholder='Email'
          value={email}
          onChange={onchange}
          required
        />
        <Input
          type='password'
          name='password'
          placeholder='Password'
          value={password}
          onChange={onchange}
          required
        />
        <Input type='submit' value='Login' />
      </Form>
      {error !== null && (
        <Errors>
          {!Array.isArray(error) ? (
            <ErrorButton type='button' onClick={() => clearErrors()}>
              {error}
            </ErrorButton>
          ) : (
            error.map(err => (
              <ErrorButton
                key={err.msg}
                type='button'
                onClick={() => clearErrors()}
              >
                {err.msg}
              </ErrorButton>
            ))
          )}
        </Errors>
      )}
      <QuestionText>Dont have an accout? &nbsp;<Link to='/register'>Sign Up</Link>
      </QuestionText>
    </Container>
  );
};

Login.propTypes = {
  history: PropTypes.object
};

export default Login;
