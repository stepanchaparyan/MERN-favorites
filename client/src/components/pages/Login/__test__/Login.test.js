import React, { useContext } from 'react';
import { renderSnapshot } from '../../../../utils/tests';
import { shallow } from 'enzyme';
import 'jest-styled-components';
import Login from '../Login.js';
import { Input, Form, ErrorButton } from '../LoginStyled';

jest.mock('react-intl', () => ({
  useIntl: jest.fn(() => ({ formatMessage: m => m.defaultMessage })),
  defineMessages: i => i
}));

jest.mock('react', () => {
  const ActualReact = jest.requireActual('react');
  return {
    ...ActualReact,
    useEffect: jest.fn().mockImplementation(f => f()),
    useContext: jest.fn(() => ({
      isAuthencated: false,
      user: { name: 'testName' },
      error: ['a', 'b'],
      login: jest.fn(),
      clearErrors: jest.fn()
    }))
  };
});

describe('Login component test with Enzyme', () => {
  test('renders snapshot', () => {
    renderSnapshot(Login);
  });

  test('should click on Input elements', () => {
    const component = shallow(<Login />);
    const inputs = component.find(Input);
    inputs.forEach(input => {
      input.simulate('change', { target: { value: 'target' } });
    });
  });

  test('should click on Form elements', () => {
    const component = shallow(<Login />);
    const form = component.find(Form);
    const preventDefault = jest.fn();
    form.simulate('submit', { preventDefault });
    expect(preventDefault).toHaveBeenCalled();
  });

  test('should click on errorButton elements', () => {
    const component = shallow(<Login />);
    const errorButtons = component.find(ErrorButton);
    errorButtons.forEach(errorButton => {
      errorButton.simulate('click');
    });
  });

  test('should click on errorButton elements when error is not an array', () => {
    useContext.mockImplementation(
      jest.fn(() => ({
        isAuthencated: false,
        user: { name: 'testName' },
        error: 'a',
        login: jest.fn(),
        clearErrors: jest.fn()
      }))
    );
    const component = shallow(<Login />);
    const errorButtons = component.find(ErrorButton);
    errorButtons.forEach(errorButton => {
      errorButton.simulate('click');
    });
  });
});
