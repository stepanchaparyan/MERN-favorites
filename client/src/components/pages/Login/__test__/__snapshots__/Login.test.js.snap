// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component test with Enzyme renders snapshot 1`] = `
<styled.div>
  <styled.div>
    LogIn
  </styled.div>
  <styled.form
    onSubmit={[Function]}
  >
    <styled.input
      name="email"
      onChange={[Function]}
      placeholder="Email"
      required={true}
      type="email"
      value=""
    />
    <styled.input
      name="password"
      onChange={[Function]}
      placeholder="Password"
      required={true}
      type="password"
      value=""
    />
    <styled.button>
      LogIn
    </styled.button>
  </styled.form>
  <styled.div>
    <styled.button
      onClick={[Function]}
      type="button"
    />
    <styled.button
      onClick={[Function]}
      type="button"
    />
  </styled.div>
  <styled.div>
    Dont have an accout?
     Â 
    <Link
      to="/register"
    >
      Sign Up
    </Link>
  </styled.div>
</styled.div>
`;
